cli/npm/managed/tarball.rs

```diff
@@ -16,19 +18,50 @@ use flate2::read::GzDecoder;
 use tar::Archive;
 use tar::EntryType;
 
-use super::cache::with_folder_sync_lock;
+use crate::util::path::get_atomic_dir_path;
 
 pub fn verify_and_extract_tarball(
-  package: &PackageNv,
+  package_nv: &PackageNv,
   data: &[u8],
   dist_info: &NpmPackageVersionDistInfo,
   output_folder: &Path,
 ) -> Result<(), AnyError> {
-  verify_tarball_integrity(package, data, &dist_info.integrity())?;
+  verify_tarball_integrity(package_nv, data, &dist_info.integrity())?;
+
+  let temp_dir = get_atomic_dir_path(output_folder);
+  extract_tarball(data, &temp_dir)?;
+  rename_with_retries(&temp_dir, output_folder)
+    .map_err(AnyError::from)
+    .context("Failed moving extracted tarball to final destination.")
+}
+
+fn rename_with_retries(
+  temp_dir: &Path,
+  output_folder: &Path,
+) -> Result<(), std::io::Error> {
+  let mut count = 0;
+  // renaming might be flaky if a lot of processes are trying
+  // to do this, so retry a few times
+  loop {
+    match fs::rename(temp_dir, output_folder) {
+      Ok(_) => return Ok(()),
+      Err(err) if err.kind() == ErrorKind::AlreadyExists => {
+        // another process copied here, just cleanup
+        let _ = fs::remove_dir_all(temp_dir);
+      }
+      Err(err) => {
+        count += 1;
+        if count >= 5 {
```

Comment:
I was getting failures at 3 on my machine, so I increased this to 5

Reply:
Actually, it seems it still failed in some cases, but much rarer now.

Reply:
It was rarer because my cache was being populated over time 😅

I have an actual fix for this for this coming soon. Just testing it out.